File: Other/Review/Phase 6/Review4/Part1/prim/prim.py
from typing import List


def prim(edges: List[List[int]], points: int) -> tuple[int, List[List[int]]]:
    mst: List[List[int]] = []
    cost = 0

    # init
    count = points - 1
    visited = [False] * points
    queue: List[List[int]] = []

    for item in edges:
        if item[0] == 0 or item[1] == 0:
            queue.append(item)
    visited[0] = True

    while count > 0:
        queue.sort(key=lambda x: x[2])

        v, w, weight = queue.pop(0)

        if not visited[v] or not visited[w]:
            mst.append([v, w])
            cost += weight
            count -= 1

            nextNode = v if not visited[v] else w
            visited[nextNode] = True
            for item in edges:
                if (item[0] == nextNode and not visited[item[1]]) or (
                    item[1] == nextNode and not visited[item[0]]
                ):
                    queue.append(item)

    return cost, mst


# test
res = prim(
    [
        [0, 1, 4],
        [0, 2, 13],
        [0, 3, 7],
        [0, 4, 7],
        [1, 2, 9],
        [1, 3, 3],
        [1, 4, 7],
        [2, 3, 10],
        [2, 4, 14],
        [3, 4, 4],
    ],
    5,
)
print(res)

File: Data Structure & Algorithm/Review/Stage 5/Review 2/prim/prim.go
package main

import (
	"fmt"
	"sort"
)

func prim(graph [][]int, vertices int) (int, [][]int) {
	mst := make([][]int, 0, vertices-1)
	cost := 0

	// init
	count := vertices - 1
	queue := make([][]int, 0, vertices-1)
	visited := make([]bool, vertices, vertices)
	for i := 0; i < vertices; i++ {
		visited[i] = false
	}

	for _, val := range graph {
		if val[0] == 0 || val[1] == 0 {
			queue = append(queue, val)
		}
	}
	visited[0] = true

	// loop
	for count > 0 {
		// sort
		sort.Slice(queue, func(i, j int) bool {
			return queue[i][2] < queue[j][2]
		})

		v, w, weight := queue[0][0], queue[0][1], queue[0][2]
		queue = queue[1:]

		if !visited[v] || !visited[w] {
			cost += weight
			mst = append(mst, []int{v, w})
			count -= 1

			var next_node int
			if visited[v] == false {
				next_node = v
			} else {
				next_node = w
			}
			visited[next_node] = true

			for _, val := range graph {
				if (val[0] == next_node && visited[val[1]] != true) || (val[1] == next_node && visited[val[0]] != true) {
					queue = append(queue, val)
				}
			}
		}
	}

	return cost, mst
}

// test
func main() {
	cost, mst := prim(
		[][]int{
			{0, 1, 4},
			{0, 2, 13},
			{0, 3, 7},
			{0, 4, 7},
			{1, 2, 9},
			{1, 3, 3},
			{1, 4, 7},
			{2, 3, 10},
			{2, 4, 14},
			{3, 4, 4},
		},
		5,
	)
	fmt.Println(cost, mst)
}
